configobject/configsync/ha.go

```diff
@@ -0,0 +1,122 @@
+package configsync
+
+import (
+	"github.com/Icinga/icingadb/ha"
+	"github.com/Icinga/icingadb/supervisor"
+	"github.com/go-redis/redis/v7"
+	log "github.com/sirupsen/logrus"
+	"sync"
+	"time"
+)
+
+const (
+	Notify_StartSync = iota
+	Notify_StopSync
+)
+
+type ConfigSyncHA struct {
+	super                      *supervisor.Supervisor
+	done                       chan struct{}
+	chHA                       <-chan ha.State
+	notificationListeners      map[string][]chan int
+	notificationListenersMutex sync.Mutex
+	lastEventId                string
+	haIsActive                 bool
+}
+
+func NewConfigSyncHA(super *supervisor.Supervisor, chHA <-chan ha.State) *ConfigSyncHA {
+	ha := &ConfigSyncHA{
+		super:                      super,
+		done:                       make(chan struct{}),
+		chHA:                       chHA,
+		notificationListeners:      map[string][]chan int{},
+		notificationListenersMutex: sync.Mutex{},
+		lastEventId:                "0-0",
+		haIsActive:                 false,
+	}
+
+	return ha
+}
+
+func (h *ConfigSyncHA) Start() {
+	go h.run()
+}
+
+func (h *ConfigSyncHA) run() {
+	every1s := time.NewTicker(time.Second)
+
+loop:
+	for {
+		select {
+		case <-h.done:
+			log.Info("received done signal, shutting down")
+			break loop
+		case <-every1s.C:
+			h.runEventListener()
+		}
+	}
+}
+
+func (h *ConfigSyncHA) Stop() {
+	close(h.done)
+}
+
+func (h *ConfigSyncHA) runEventListener() {
+	select {
+	case newState := <-h.chHA:
+		newIsActive := newState == ha.StateActive
+		h.haIsActive = newIsActive
+		if !newIsActive {
+			h.lastEventId = "0-0"
+			h.notifyNotificationListener("*", Notify_StopSync)
+		}
+	default: // don't block if there is no change
+	}
+
+	if !h.haIsActive {
+		return
```

Comment:
Why not to select `<-h.done` and `<-h.chHA`?

Reply:
So block until one of these happens? The code following is supposed to happen every second, see the ticker in `run()`.

Reply:
And why exactly?

Reply:
Because that's what the code was previously doing and I didn't rewrite this completely. Most parts of this file are just moved from the old implementation: https://github.com/Icinga/icingadb/blob/88deca05b73f95f86f1bc91fd1bb83ed42494be6/ha/ha.go#L301-L339

Reply:
Fine. Other construction area(tm).